fvm/evm/offchain/sync/replay.go

```diff
@@ -0,0 +1,250 @@
+package sync
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/onflow/atree"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	gethTypes "github.com/onflow/go-ethereum/core/types"
+	gethTracer "github.com/onflow/go-ethereum/eth/tracers"
+	gethTrie "github.com/onflow/go-ethereum/trie"
+
+	"github.com/onflow/flow-go/fvm/evm/emulator"
+	"github.com/onflow/flow-go/fvm/evm/events"
+	"github.com/onflow/flow-go/fvm/evm/offchain/blocks"
+	"github.com/onflow/flow-go/fvm/evm/offchain/storage"
+	"github.com/onflow/flow-go/fvm/evm/precompiles"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+var emptyChecksum = [types.ChecksumLength]byte{0, 0, 0, 0}
+
+// ReplayBlockExecution re-executes transactions of a block using the
+// events emitted when transactions where executed.
+// it updates the state of the given ledger and uses the trace
+func ReplayBlockExecution(
+	chainID flow.ChainID,
+	rootAddr flow.Address,
+	snapshot types.BackendStorageSnapshot,
+	tracer *gethTracer.Tracer,
+	transactionEvents []events.TransactionEventPayload,
+	blockEvent *events.BlockEventPayload,
+	validateResults bool,
+) (types.ReplayResults, error) {
+
+	// create storage
+	storage := storage.NewEphemeralStorage(storage.NewReadOnlyStorage(snapshot))
+
+	// prepare blocks
+	blks, err := blocks.NewBlocks(chainID, rootAddr, storage)
+	if err != nil {
+		return nil, err
+	}
+	// push the new block meta
+	// it should be done before execution so block context creation
+	// can be done properly
+	err = blks.PushBlockMeta(
+		blocks.NewMeta(
+			blockEvent.Height,
+			blockEvent.Timestamp,
+			blockEvent.PrevRandao,
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a base block context for all transactions
+	// tx related context values will be replaced during execution
+	ctx, err := CreateBlockContext(chainID, blks, tracer)
+	if err != nil {
+		return nil, err
+	}
+
+	gasConsumedSoFar := uint64(0)
+	txHashes := make(types.TransactionHashes, len(transactionEvents))
+	for idx, tx := range transactionEvents {
+		err = replayTransactionExecution(
+			rootAddr,
+			ctx,
+			uint(idx),
+			gasConsumedSoFar,
+			storage,
+			&tx,
+			validateResults,
+		)
+		if err != nil {
+			return nil, err
+		}
+		gasConsumedSoFar += tx.GasConsumed
+		txHashes[idx] = tx.Hash
+	}
+
+	if validateResults {
+		// check transaction inclusion
+		txHashRoot := gethTypes.DeriveSha(txHashes, gethTrie.NewStackTrie(nil))
+		if txHashRoot != blockEvent.TransactionHashRoot {
```

Comment:
The `blockEvent` is of type `*events.BlockEventPayload`, so we might want to add some `nil` checks to nil pointer dereference errors.

Reply:
gonna add a check on the start of the method.